### PR TITLE
Major update

### PR DESCRIPTION
**Pull request by reidmv**
This PR is a proposed refactor and major version update. This would be released as a new major version.
- Switch to Puppet 4
  - Use type features for validation, eliminating dependency on puppetlabs/stdlib.
  - Use iteration in dotnet::install::package utility type
  - Use new `$::os` fact
- Switch from puppet/download_file to lwf/remote_file. Per commit message in 4db8e15, primary benefit is cleaner implementation allowing for more reliable notify/subscribe. Especially when reboots are subscribed to a dotnet resource this can be important.
- Switch from exec-based package installation to package resource.
  - Makes it easier to see what dotnet is doing in the graph
  - Makes the code cleaner

### CODE REVIEW

**Comment by liamjbennett** [comment]
I really like some of these changes so thank you.
I want to give this a more detailed review but initially I have some questions:
1) Why the use of the remote_file dependency Why this over download_file, archive or staging?
2) Why the switch to Puppet 4 already?

**Comment by reidmv** [comment]
Hi Liam,

Excellent questions! I figured I'd start with the PR as far forward as I already took it for my project use, and if necessary whittle things down based on the code review. With that background, here is my rationale for using remote_file and Puppet 4:

1) Archive has client-side dependencies and so is what I would consider heavyweight; in addition to that there's no need to extract anything here. Staging has the same problem as download_file. That problem for me is that I'm setting up a reboot resource subscribed to e.g. `Dotnet['.NET Framework 4.0']`, and during development I ran into a situation where the source EXE didn't need to be downloaded nor .NET installed, but a private administrative resource (this [file](https://github.com/puppet-community/puppet-download_file/blob/master/manifests/init.pp#L76-L81)) in download_file got changed, thus triggering a reboot. Remote_file is a native type that doesn't have administrative fragments that could generate extra events. That reduces the surface area for unnecessary resource changes, thus lowering the chances of accidentally notifying something (like a reboot) down the line. Since I was proposing a major revision anyway it made sense to suggest it now. There's also the nifty side-benefit of remote_file being platform agnostic. 

2) The simple answer is I mostly write Puppet 4 / future parser code these days. When thinking about how to deal with specifying multiple versions of .NET which don't technically co-exist I ended up using iteration [here](https://github.com/reidmv/puppet-dotnet/blob/update/manifests/install/package.pp#L32-L46). With that toehold I went ahead and added type checking as well, allowing the removal of the stdlib dependency. As long as I was removing dependencies I figured using the `$::os` fact would be a good idea too instead of the windowsfacts-provided `$::operatingsystemversion` so that the dependencies could be slimmed down that much more as well. On `$::os` I can have high confidence that the version of Facter shipping with Puppet 4 provides it due to the way Puppet Labs is packaging universal single packages for most platforms starting with Puppet 4.

The switch to Puppet 4 is of course what would necessitate making this a new major version.

**Comment by nibalizer** [comment]
From a code mgmt perspective I suggest this. Ideally it could/would apply to every module in p-c.

Cutover:
1) Branch currrent master to 1.x
2) Merge this PR to master
3) Add a Box such as https://github.com/puppetlabs/puppetlabs-stdlib#version-compatibility to the README explaining which versions of the module are compatible with which puppet versions. 
4) Release 2.0.0 of the module to the forge
5) Release the tip of 1.x to the forge as 1.1.whatever if it has been long enough since the last 1.x release.

Ongoing: 
* New features should land on 2.x.
* Bugfixes and backports can land on 1.x, but only when things seem super broken and important. We should strive to help the consumers move to puppet4.
* Releases of the 2.x line can continue as normal.
* Releases of the 1.x.z line (probably .z bugfix releases) can be released at any time. I'm personally okay with releasing after every PR since there shouldn't be many of them and we'd only get that PR if someone was in a rough spot to begin with.
* At some point in the future we'll depreciate all puppet3 compatibility.

**Comment by liamjbennett** [comment]
Ok, I am happy for this change to go ahead and I will do all the required stuff to release a new major version on this.

@reidmv can you fix the tests first on this? Thanks.

**Comment by reidmv** [comment]
I can, but not soon. I'm pretty swamped until the end of October. If I get a chance to slip in more work on this before then I will. I'll post an update by October 27.

**Comment by nibalizer** [comment]
I'm trying to get the tests passing in #10

**Comment by reidmv** [comment]
@liamjbennett @nibalizer tests cleaned up and PR rebased. Ready for review.

File: manifests/init.pp [hunk 1]
```diff
@@ -31,54 +31,77 @@
...
+  $windows_version = $::os['release']['full']
+
   case $version {
     '3.5': {
-      case $::operatingsystemversion {
-        /^Windows.Server.(2008|2012).?(R2)?.*/: { $type = 'feature' }
-        /^Windows (XP|Vista|7|8|8.1).*/: { $type = 'package' }
-        default: { $type = 'err' err("dotnet ${version} is not support on this version of windows") }
+      case $windows_version {
+        /^2008/, /^2012/:  { $type = 'feature' }
```
Comment (vincent-aguillaume) [inline manifests/init.pp]: On Windows server 2008 (non R2), dotnet 3.5 is a package.

Comment (reidmv) [inline manifests/init.pp]: Thanks @vincent-aguillaume - I've added a commit to fix the error.

**Comment by reidmv** [comment]
@liamjbennett @nibalizer bump

**Comment by jyaworski** [comment]
Hello:

What's the status of this PR? Can it be rebased and squashed?

**Comment by reidmv** [comment]
Rebased and squashed.

**Comment by liamjbennett** [comment]
@reidmv can you update to rspec-puppet 2.3.2 and try again?

**Comment by reidmv** [comment]
@liamjbennett seems to work! Needed to do style cleanup for rubocop to get that to pass, but the Travis build is green now.

**Comment by jyaworski** [comment]
Can this be squashed?

**Comment by reidmv** [comment]
@jyaworski squashed.

**Comment by reidmv** [comment]
Now modified (and squashed) to comply with strict variables.

File: Gemfile [hunk 2]
```diff
@@ -16,7 +16,7 @@ source 'https://rubygems.org'
 
 group :test do
   gem 'rake'
-  gem 'puppet', ENV['PUPPET_VERSION'] || '~> 3.8.0'
+  gem 'puppet', ENV['PUPPET_VERSION'] || '~> 4.0'
```
Comment (jyaworski) [inline Gemfile]: Why this change?

Comment (reidmv) [inline Gemfile]: @jyaworski This PR is proposing a major version update, and uses Puppet 4 language features. Therefore it is not compatible with any version of Puppet <4.0. I wrote a little more about that [here](https://github.com/voxpupuli/puppet-dotnet/pull/9#issuecomment-142127835).

Comment (jyaworski) [inline Gemfile]: The Gemfile is (or will be) managed by modulesync.

https://github.com/voxpupuli/modulesync_config

Please raise a PR over there about it, or this will get overwritten the next time we sync.

Comment (reidmv) [inline Gemfile]: I've added an issue. voxpupuli/modulesync_config#102.

**Comment by reidmv** [comment]
@jyaworski PR refactored to take modulesync into account.

**Comment by bastelfreak** [comment]
Hi @reidmv @liamjbennett, how is it planned to process here? @liamjbennett do you want to do the release (we should do another modulesync before that)?

**Comment by liamjbennett** [comment]
Yes I want this to be merged into master. I not not in a position to do the release myself at the moment but I someone else can prepare it on my behalf it would be greatly appreciated.

**Comment by bastelfreak** [comment]
@liamjbennett can we talk about the further steps on IRC? I'm currently doing a modulesync in https://github.com/voxpupuli/puppet-dotnet/pull/20, I would like to release the module in the next week.

**Comment by vinzent** [comment]
closing because of no progress since over a year
